server/service/microsoft_mdm.go

```diff
@@ -541,48 +374,141 @@ func NewSoapResponse(payload interface{}, relatesTo string) (SoapResponse, error
+func mdmMicrosoftPolicyEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (errorer, error) {
+	req := request.(*SoapRequestContainer).Data
+
+	// Checking first if GetPolicies message is valid and returning error if this is not the case
+	if err := req.IsValidGetPolicyMsg(); err != nil {
+		soapFault := svc.GetAuthorizedSoapFault(ctx, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err)
+		return getSoapResponseFault(req.GetMessageID(), soapFault), nil
+	}
+
+	// Binary security token should be extracted to ensure this is a valid call
+	binSecTokenData, err := req.GetBinarySecurityToken()
+	if err != nil {
+		soapFault := svc.GetAuthorizedSoapFault(ctx, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err)
+		return getSoapResponseFault(req.GetMessageID(), soapFault), nil
+	}
+
+	// Getting the GetPoliciesResponse message
+	discoveryMessage, err := svc.GetMDMMicrosoftPolicyResponse(ctx, binSecTokenData)
+	if err != nil {
+		soapFault := svc.GetAuthorizedSoapFault(ctx, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err)
+		return getSoapResponseFault(req.GetMessageID(), soapFault), nil
 	}
 
 	// Embedding the DiscoveryResponse message inside of a SoapResponse
-	response, err := NewSoapResponse(discoveryMessage, req.getMessageID())
+	response, err := NewSoapResponse(discoveryMessage, req.GetMessageID())
 	if err != nil {
 		soapFault := svc.GetAuthorizedSoapFault(ctx, mdm.SoapErrorMessageFormat, mdm_types.MDEDiscovery, err)
-		return getSoapResponseFault(req.getMessageID(), soapFault), nil
+		return getSoapResponseFault(req.GetMessageID(), soapFault), nil
+	}
+
+	return SoapResponseContainer{
+		Data: &response,
+		Err:  nil,
+	}, nil
+}
+
+// validateBinarySecurityToken checks if the provided token is valid
+func validateBinarySecurityToken(ctx context.Context, encodedBinarySecToken string, svc fleet.Service) error {
+	if len(encodedBinarySecToken) == 0 {
+		return errors.New("binarySecurityTokenValidation: encoded token is invalid")
+	}
+
+	// Getting the Binary Security Token Payload
+	binSecToken, err := NewBinarySecurityTokenPayload(encodedBinarySecToken)
+	if err != nil {
+		return fmt.Errorf("binarySecurityTokenValidation: token creation error %v", err)
+	}
+
+	// Validating the Binary Security Token Payload
+	err = binSecToken.IsValidToken()
+	if err != nil {
+		return fmt.Errorf("binarySecurityTokenValidation: invalid token data %v", err)
 	}
 
-	return response, nil
+	// Validating the Binary Security Token Type used on Programmatic Enrollments
+	if binSecToken.Type == mdm_types.WindowsMDMProgrammaticEnrollmentType {
+		host, err := svc.HostByIdentifier(ctx, binSecToken.Payload.HostUUID, fleet.HostDetailOptions{})
+		if err != nil {
+			return fmt.Errorf("binarySecurityTokenValidation: host data cannot be found %v", err)
+		}
+
+		// Extra checks could be done here to ensure that host is ready for MDM enrollment
+		if host.Platform != "Windows" {
```

Comment:
:+1: , I think we should re-check that `IsElegibleForWindowsMDMEnrollment` is still true for that host (e.g. it could've been enrolled in a third-party since the notification got sent to enroll, or it could be a script-based API call without having received the notification, etc.).

Reply:
Although we have to make sure this is the right place to make that check, e.g. `validateBinarySecurityToken` is called from `GetMDMMicrosoftPolicyResponse`, is the host still unenrolled at that point or would the osquery that checks for mdm enrollment return an "enrolled" state at that point?

Reply:
@marcosd4h  My understanding from the protocol flow in the diagram here: https://github.com/fleetdm/fleet/issues/12263 is that at the Get Policies stage, the host is still unenrolled so it would be ok to call `IsElegible...` here, what do you think?